### PR TITLE
Enforce unique bucket names with an extra Redis database

### PR DESCRIPTION
**Pull request by fvennetier**
A.K.A. the "bucket DB".

Besides making bucket names unique across the namespace, this database allows anonymous requests. Indeed the database is used to find which account the requested bucket belongs to, in order to check ACLs.

Note: this PR must be rebased and merged after #2.

### CODE REVIEW

File: etc/proxy-server.conf-sample [hunk 1]
```diff
@@ -580,6 +580,32 @@ use = egg:swift#s3api
 # You can override the default log routing for this filter here:
 # log_name = s3api
 
+# Enable the Redis-based bucket database. This will make the bucket names
```
Comment (murlock) [inline etc/proxy-server.conf-sample:583]: IMO, we should reuse our new format 
```
connection = redis+sentinel://10.0.1.24:6012,10.0.1.27:6012,10.0.1.25:6012?sentinel_name=bucket_db
```

File: swift/common/middleware/s3api/bucket_db.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+# limitations under the License.
+
+import time
+
+from swift.common.utils import config_true_value
+
+from oio.common.redis_conn import RedisConnection
+
+
+class DummyBucketDb(object):
```
Comment (murlock) [inline swift/common/middleware/s3api/bucket_db.py:23]: We should move this class within `tests/`

**Review by murlock — APPROVED** [review]

File: swift/common/middleware/s3api/bucket_db.py (hunk 2 above)
Comment (fvennetier) [reply swift/common/middleware/s3api/bucket_db.py:23]: I'm not sure about this. I like the fact that `get_bucket_db` does the instantiation work.

File: etc/proxy-server.conf-sample (hunk 1 above)
Comment (fvennetier) [reply etc/proxy-server.conf-sample:583]: Ok, done.

**Review by AymericDu — APPROVED** [review]

**Pull request merged**